client-sdk/ts-web/rt/src/mrae.ts

```diff
@@ -0,0 +1,53 @@
+import * as oasis from '@oasisprotocol/client';
+import * as ed from '@noble/ed25519';
+import {sha512_256} from 'js-sha512';
+
+var deoxysii = require('deoxysii');
+
+const BoxKDFTweak = 'MRAE_Box_Deoxys-II-256-128';
+
+/**
+ * deriveSymmetricKey derives a MRAE AEAD symmetric key suitable for use with the asymmetric
+ * box primitives from the provided X25519 public and private keys.
+ */
+
+export function deriveSymmetricKey(publicKey: Uint8Array, privateKey: Uint8Array): Uint8Array {
+    const pmk = ed.curve25519.scalarMult(oasis.misc.toHex(privateKey), oasis.misc.toHex(publicKey));
+    var kdf = sha512_256.hmac.create(BoxKDFTweak);
+    kdf.update(pmk);
+    return oasis.misc.fromHex(kdf.hex());
+}
+
+/**
+ * boxSeal boxes ("seals") the provided additional data and plaintext via
+ * Deoxys-II-256-128 using a symmetric key derived from the provided
+ * X25519 public and private keys.
+ */
+export function boxSeal(
+    nonce: Uint8Array,
+    plainText: Uint8Array,
+    associateData: Uint8Array,
+    publicKey: Uint8Array,
+    privateKey: Uint8Array,
```

Comment:
Can you use something like
```ts
type Nonce = Opaque<Uint8Array>;
type PublicKey = Opaque<Uint8Array>;
type SecretKey = Opaque<Uint8Array>;
```
or take named args via an object so that it's harder to mix up these five positional args of the same type?

[docs for `Opaque`](https://github.com/sindresorhus/type-fest/blob/HEAD/source/opaque.d.ts)

Reply:
this whole sdk is built to operate directly on types that show up from grpc. it's all Uint8Array. keys, hashes, encoded structures, quantities, IDs, addresses, aaaaaaaaaaaa

https://github.com/oasisprotocol/oasis-sdk/tree/a0da10b705b91aaf902dfa3aa0562af25a32f38c/client-sdk/ts-web/core#design-notes design note number 2